File: sim/optimizer/specs.go
package optimizer

import (
	"slices"

	"github.com/wowsims/classic/sim/core/proto"
)

// SpecDef is the Go twin of ui/app/specs.ts: what the Forever app sims for a spec and what
// the class can equip at a level. Keep the two in step (TestSpecsMatchApp checks the keys,
// rotations and proficiency lists against specs.ts).
type SpecDef struct {
	Key, Label string
	Class      proto.Class
	Role       string // melee | ranged | caster
	Rotations  []RotationDef
	Spec       func(level int32) interface{} // a proto.Player_<Spec> for core.WithSpec
	Armor      func(level int32) proto.ArmorType
	Weapons    []proto.WeaponType
	TwoHand    bool
	DualWield  func(level int32) bool
	Ranged     []proto.RangedWeaponType
	Shield     bool
}

// RotationDef is an APL file under ui/ (Path without the .apl.json suffix).
type RotationDef struct {
	Label, Path string
}

const (
	wAxe     = proto.WeaponType_WeaponTypeAxe
	wDagger  = proto.WeaponType_WeaponTypeDagger
	wFist    = proto.WeaponType_WeaponTypeFist
	wMace    = proto.WeaponType_WeaponTypeMace
	wOffHand = proto.WeaponType_WeaponTypeOffHand
	wPolearm = proto.WeaponType_WeaponTypePolearm
	wShield  = proto.WeaponType_WeaponTypeShield
	wStaff   = proto.WeaponType_WeaponTypeStaff
	wSword   = proto.WeaponType_WeaponTypeSword

	rBow      = proto.RangedWeaponType_RangedWeaponTypeBow
	rCrossbow = proto.RangedWeaponType_RangedWeaponTypeCrossbow
	rGun      = proto.RangedWeaponType_RangedWeaponTypeGun
	rIdol     = proto.RangedWeaponType_RangedWeaponTypeIdol
	rLibram   = proto.RangedWeaponType_RangedWeaponTypeLibram
	rThrown   = proto.RangedWeaponType_RangedWeaponTypeThrown
	rTotem    = proto.RangedWeaponType_RangedWeaponTypeTotem
	rWand     = proto.RangedWeaponType_RangedWeaponTypeWand

	cloth   = proto.ArmorType_ArmorTypeCloth
	leather = proto.ArmorType_ArmorTypeLeather
	mail    = proto.ArmorType_ArmorTypeMail
	plate   = proto.ArmorType_ArmorTypePlate
)

func always(v proto.ArmorType) func(int32) proto.ArmorType { return func(int32) proto.ArmorType { return v } }
func from40(low, high proto.ArmorType) func(int32) proto.ArmorType {
	return func(l int32) proto.ArmorType {
		if l >= 40 {
			return high
		}
		return low
	}
}
func never(int32) bool     { return false }
func from20(l int32) bool  { return l >= 20 }
func alwaysDW(int32) bool  { return true }

// Specs lists the DPS specs of the Forever app, in the app's order.
var Specs = []*SpecDef{
	{
		Key: "warrior", Label: "Warrior", Class: proto.Class_ClassWarrior, Role: "melee",
		Rotations: []RotationDef{{"Leveling (any level)", "warrior/apls/leveling"}, {"Level 60 Fury/Arms", "warrior/apls/dps_no_reck"}},
		Spec: func(l int32) interface{} {
			stance := proto.WarriorStance_WarriorStanceBattle
			if l >= 30 {
				stance = proto.WarriorStance_WarriorStanceBerserker
			}
			return &proto.Player_Warrior{Warrior: &proto.Warrior{Options: &proto.Warrior_Options{
				StartingRage: 0, QueueDelay: 250, Shout: proto.WarriorShout_WarriorShoutBattle, Stance: stance}}}
		},
		Armor:   from40(mail, plate),
		Weapons: []proto.WeaponType{wAxe, wDagger, wFist, wMace, wPolearm, wStaff, wSword}, TwoHand: true, DualWield: from20,
		Ranged: []proto.RangedWeaponType{rBow, rCrossbow, rGun, rThrown}, Shield: true,
	},
	{
		Key: "rogue", Label: "Rogue", Class: proto.Class_ClassRogue, Role: "melee",
		Rotations: []RotationDef{{"Sinister Strike", "rogue/apls/leveling_sinister_strike"}, {"Backstab (dagger)", "rogue/apls/leveling_backstab"},
			{"Level 60 Sinister Strike", "rogue/apls/combat_sinister_strike"}, {"Level 60 Backstab", "rogue/apls/combat_backstab"}},
		Spec: func(int32) interface{} {
			return &proto.Player_Rogue{Rogue: &proto.Rogue{Options: &proto.RogueOptions{}}}
		},
		Armor:   always(leather),
		Weapons: []proto.WeaponType{wAxe, wDagger, wFist, wMace, wSword}, TwoHand: false, DualWield: alwaysDW,
		Ranged: []proto.RangedWeaponType{rBow, rCrossbow, rGun, rThrown}, Shield: false,
	},
	{
		Key: "hunter", Label: "Hunter", Class: proto.Class_ClassHunter, Role: "ranged",
		Rotations: []RotationDef{{"Leveling (any level)", "hunter/apls/leveling"}, {"Level 60", "hunter/apls/p1"}},
		Spec: func(int32) interface{} {
			return &proto.Player_Hunter{Hunter: &proto.Hunter{Options: &proto.Hunter_Options{
				PetType: proto.Hunter_Options_Cat, PetUptime: 1, PetAttackSpeed: proto.Hunter_Options_OneTwo}}}
		},
		Armor:   from40(leather, mail),
		Weapons: []proto.WeaponType{wAxe, wDagger, wFist, wPolearm, wStaff, wSword}, TwoHand: true, DualWield: from20,
		Ranged: []proto.RangedWeaponType{rBow, rCrossbow, rGun}, Shield: false,
	},
	{
		Key: "mage", Label: "Mage", Class: proto.Class_ClassMage, Role: "caster",
		Rotations: []RotationDef{{"Fireball", "mage/apls/leveling_fireball"}, {"Frostbolt", "mage/apls/leveling_frostbolt"}, {"Level 60 Fire", "mage/apls/p1"}},
		Spec: func(l int32) interface{} {
			armor := proto.Mage_Options_NoArmor
			if l >= 34 {
				armor = proto.Mage_Options_MageArmor
			}
			return &proto.Player_Mage{Mage: &proto.Mage{Options: &proto.Mage_Options{Armor: armor}}}
		},
		Armor:   always(cloth),
		Weapons: []proto.WeaponType{wDagger, wStaff, wSword, wOffHand}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rWand}, Shield: false,
	},
	{
		Key: "warlock", Label: "Warlock", Class: proto.Class_ClassWarlock, Role: "caster",
		Rotations: []RotationDef{{"Default", "warlock/apls/rotation"}},
		Spec: func(l int32) interface{} {
			summon := proto.WarlockOptions_Imp
			if l >= 20 {
				summon = proto.WarlockOptions_Succubus
			}
			return &proto.Player_Warlock{Warlock: &proto.Warlock{Options: &proto.WarlockOptions{Armor: proto.WarlockOptions_DemonArmor, Summon: summon}}}
		},
		Armor:   always(cloth),
		Weapons: []proto.WeaponType{wDagger, wStaff, wSword, wOffHand}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rWand}, Shield: false,
	},
	{
		Key: "shadow_priest", Label: "Shadow Priest", Class: proto.Class_ClassPriest, Role: "caster",
		Rotations: []RotationDef{{"Default", "shadow_priest/apls/p1"}},
		Spec: func(int32) interface{} {
			return &proto.Player_ShadowPriest{ShadowPriest: &proto.ShadowPriest{Options: &proto.ShadowPriest_Options{}}}
		},
		Armor:   always(cloth),
		Weapons: []proto.WeaponType{wDagger, wMace, wStaff, wOffHand}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rWand}, Shield: false,
	},
	{
		Key: "balance_druid", Label: "Balance Druid", Class: proto.Class_ClassDruid, Role: "caster",
		Rotations: []RotationDef{{"Default", "balance_druid/apls/balance"}},
		Spec: func(int32) interface{} {
			return &proto.Player_BalanceDruid{BalanceDruid: &proto.BalanceDruid{Options: &proto.BalanceDruid_Options{}}}
		},
		Armor:   always(leather),
		Weapons: []proto.WeaponType{wDagger, wFist, wMace, wStaff, wOffHand}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rIdol}, Shield: false,
	},
	{
		Key: "feral_druid", Label: "Feral Druid", Class: proto.Class_ClassDruid, Role: "melee",
		Rotations: []RotationDef{{"Leveling (any level)", "feral_druid/apls/leveling"}, {"Level 60", "feral_druid/apls/feral"}},
		Spec: func(int32) interface{} {
			return &proto.Player_FeralDruid{FeralDruid: &proto.FeralDruid{Options: &proto.FeralDruid_Options{LatencyMs: 100}}}
		},
		Armor:   always(leather),
		Weapons: []proto.WeaponType{wDagger, wFist, wMace, wStaff, wOffHand}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rIdol}, Shield: false,
	},
	{
		Key: "retribution_paladin", Label: "Retribution Paladin", Class: proto.Class_ClassPaladin, Role: "melee",
		Rotations: []RotationDef{{"Default", "retribution_paladin/apls/basic_ret"}},
		Spec: func(l int32) interface{} {
			aura := proto.PaladinAura_NoPaladinAura
			if l >= 16 {
				aura = proto.PaladinAura_RetributionAura
			}
			return &proto.Player_RetributionPaladin{RetributionPaladin: &proto.RetributionPaladin{Options: &proto.PaladinOptions{
				Aura: aura, PrimarySeal: proto.PaladinSeal_Righteousness}}}
		},
		Armor:   from40(mail, plate),
		Weapons: []proto.WeaponType{wAxe, wMace, wPolearm, wSword}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rLibram}, Shield: true,
	},
	{
		Key: "elemental_shaman", Label: "Elemental Shaman", Class: proto.Class_ClassShaman, Role: "caster",
		Rotations: []RotationDef{{"Default", "elemental_shaman/apls/default"}},
		Spec: func(int32) interface{} {
			return &proto.Player_ElementalShaman{ElementalShaman: &proto.ElementalShaman{Options: &proto.ElementalShaman_Options{}}}
		},
		Armor:   from40(leather, mail),
		Weapons: []proto.WeaponType{wAxe, wDagger, wFist, wMace, wStaff, wOffHand}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rTotem}, Shield: true,
	},
	{
		Key: "enhancement_shaman", Label: "Enhancement Shaman", Class: proto.Class_ClassShaman, Role: "melee",
		Rotations: []RotationDef{{"Default", "enhancement_shaman/apls/default"}},
		Spec: func(int32) interface{} {
			return &proto.Player_EnhancementShaman{EnhancementShaman: &proto.EnhancementShaman{Options: &proto.EnhancementShaman_Options{}}}
		},
		Armor:   from40(leather, mail),
		Weapons: []proto.WeaponType{wAxe, wDagger, wFist, wMace, wStaff}, TwoHand: true, DualWield: never,
		Ranged: []proto.RangedWeaponType{rTotem}, Shield: true,
	},
}

// SpecByKey returns the spec definition for an app key (e.g. "warrior", "rogue").
func SpecByKey(key string) *SpecDef {
	for _, s := range Specs {
		if s.Key == key {
			return s
		}
	}
	return nil
}

func (s *SpecDef) usesWeapon(t proto.WeaponType) bool       { return slices.Contains(s.Weapons, t) }
func (s *SpecDef) usesRanged(t proto.RangedWeaponType) bool { return slices.Contains(s.Ranged, t) }

File: ui/app/specs.ts
/// <reference types="vite/client" />
// The DPS specs the Forever app offers, with what each one needs to build a request:
// level-appropriate class options, the rotation the site ships, and what the class can equip.
import { Player } from '../core/proto/api';
import { APLRotation } from '../core/proto/apl';
import { Class } from '../core/proto/common';
import { BalanceDruid, BalanceDruid_Options, FeralDruid, FeralDruid_Options } from '../core/proto/druid';
import { Hunter, Hunter_Options, Hunter_Options_PetAttackSpeed, Hunter_Options_PetType } from '../core/proto/hunter';
import { Mage, Mage_Options, Mage_Options_ArmorType } from '../core/proto/mage';
import { PaladinAura, PaladinOptions, PaladinSeal, RetributionPaladin } from '../core/proto/paladin';
import { ShadowPriest, ShadowPriest_Options } from '../core/proto/priest';
import { Rogue, RogueOptions } from '../core/proto/rogue';
import { ElementalShaman, ElementalShaman_Options, EnhancementShaman, EnhancementShaman_Options } from '../core/proto/shaman';
import { Warlock, WarlockOptions, WarlockOptions_Armor, WarlockOptions_Summon } from '../core/proto/warlock';
import { Warrior, Warrior_Options, WarriorShout, WarriorStance } from '../core/proto/warrior';

import balanceApl from '../balance_druid/apls/balance.apl.json?raw';
import elementalApl from '../elemental_shaman/apls/default.apl.json?raw';
import enhancementApl from '../enhancement_shaman/apls/default.apl.json?raw';
import feralLevelApl from '../feral_druid/apls/leveling.apl.json?raw';
import feralApl from '../feral_druid/apls/feral.apl.json?raw';
import hunterLevelApl from '../hunter/apls/leveling.apl.json?raw';
import hunterApl from '../hunter/apls/p1.apl.json?raw';
import mageFireballApl from '../mage/apls/leveling_fireball.apl.json?raw';
import mageFrostboltApl from '../mage/apls/leveling_frostbolt.apl.json?raw';
import mageApl from '../mage/apls/p1.apl.json?raw';
import retApl from '../retribution_paladin/apls/basic_ret.apl.json?raw';
import rogueLevelBsApl from '../rogue/apls/leveling_backstab.apl.json?raw';
import rogueLevelApl from '../rogue/apls/leveling_sinister_strike.apl.json?raw';
import rogueBackstabApl from '../rogue/apls/combat_backstab.apl.json?raw';
import rogueApl from '../rogue/apls/combat_sinister_strike.apl.json?raw';
import shadowApl from '../shadow_priest/apls/p1.apl.json?raw';
import warlockApl from '../warlock/apls/rotation.apl.json?raw';
import warriorLevelApl from '../warrior/apls/leveling.apl.json?raw';
import warriorApl from '../warrior/apls/dps_no_reck.apl.json?raw';

export const enum Armor { Cloth = 1, Leather = 2, Mail = 3, Plate = 4 }
// proto WeaponType / RangedWeaponType values.
const W = { axe: 1, dagger: 2, fist: 3, mace: 4, offhand: 5, polearm: 6, shield: 7, staff: 8, sword: 9 };
const R = { bow: 1, crossbow: 2, gun: 3, idol: 4, libram: 5, thrown: 6, totem: 7, wand: 8 };

export interface SpecDef {
	key: string;
	label: string;
	className: string;
	cls: Class;
	rotations: Array<{ label: string; json: string }>;
	spec: (level: number) => Player['spec'];
	armor: (level: number) => Armor;
	weapons: number[];
	twoHand: boolean;
	dualWield: (level: number) => boolean;
	ranged: number[];
	shield: boolean;
	role: 'melee' | 'ranged' | 'caster';
}

export const SPECS: SpecDef[] = [
	{
		key: 'warrior', label: 'Warrior', className: 'Warrior', cls: Class.ClassWarrior, role: 'melee',
		rotations: [{ label: 'Leveling (any level)', json: warriorLevelApl }, { label: 'Level 60 Fury/Arms', json: warriorApl }],
		spec: level => ({ oneofKind: 'warrior', warrior: Warrior.create({ options: Warrior_Options.create({
			startingRage: 0, queueDelay: 250, shout: WarriorShout.WarriorShoutBattle,
			stance: level >= 30 ? WarriorStance.WarriorStanceBerserker : WarriorStance.WarriorStanceBattle }) }) }),
		armor: level => (level >= 40 ? Armor.Plate : Armor.Mail),
		weapons: [W.axe, W.dagger, W.fist, W.mace, W.polearm, W.staff, W.sword], twoHand: true, dualWield: level => level >= 20,
		ranged: [R.bow, R.crossbow, R.gun, R.thrown], shield: true,
	},
	{
		key: 'rogue', label: 'Rogue', className: 'Rogue', cls: Class.ClassRogue, role: 'melee',
		rotations: [{ label: 'Sinister Strike', json: rogueLevelApl }, { label: 'Backstab (dagger)', json: rogueLevelBsApl }, { label: 'Level 60 Sinister Strike', json: rogueApl }, { label: 'Level 60 Backstab', json: rogueBackstabApl }],
		spec: () => ({ oneofKind: 'rogue', rogue: Rogue.create({ options: RogueOptions.create({}) }) }),
		armor: () => Armor.Leather,
		// Forever lets rogues use one-handed axes.
		weapons: [W.axe, W.dagger, W.fist, W.mace, W.sword], twoHand: false, dualWield: () => true,
		ranged: [R.bow, R.crossbow, R.gun, R.thrown], shield: false,
	},
	{
		key: 'hunter', label: 'Hunter', className: 'Hunter', cls: Class.ClassHunter, role: 'ranged',
		rotations: [{ label: 'Leveling (any level)', json: hunterLevelApl }, { label: 'Level 60', json: hunterApl }],
		spec: () => ({ oneofKind: 'hunter', hunter: Hunter.create({ options: Hunter_Options.create({
			petType: Hunter_Options_PetType.Cat, petUptime: 1, petAttackSpeed: Hunter_Options_PetAttackSpeed.OneTwo }) }) }),
		armor: level => (level >= 40 ? Armor.Mail : Armor.Leather),
		weapons: [W.axe, W.dagger, W.fist, W.polearm, W.staff, W.sword], twoHand: true, dualWield: level => level >= 20,
		ranged: [R.bow, R.crossbow, R.gun], shield: false,
	},
	{
		key: 'mage', label: 'Mage', className: 'Mage', cls: Class.ClassMage, role: 'caster',
		rotations: [{ label: 'Fireball', json: mageFireballApl }, { label: 'Frostbolt', json: mageFrostboltApl }, { label: 'Level 60 Fire', json: mageApl }],
		spec: level => ({ oneofKind: 'mage', mage: Mage.create({ options: Mage_Options.create({
			armor: level >= 34 ? Mage_Options_ArmorType.MageArmor : Mage_Options_ArmorType.NoArmor }) }) }),
		armor: () => Armor.Cloth,
		weapons: [W.dagger, W.staff, W.sword, W.offhand], twoHand: true, dualWield: () => false,
		ranged: [R.wand], shield: false,
	},
	{
		key: 'warlock', label: 'Warlock', className: 'Warlock', cls: Class.ClassWarlock, role: 'caster',
		rotations: [{ label: 'Default', json: warlockApl }],
		spec: level => ({ oneofKind: 'warlock', warlock: Warlock.create({ options: WarlockOptions.create({
			armor: WarlockOptions_Armor.DemonArmor,
			summon: level >= 20 ? WarlockOptions_Summon.Succubus : WarlockOptions_Summon.Imp }) }) }),
		armor: () => Armor.Cloth,
		weapons: [W.dagger, W.staff, W.sword, W.offhand], twoHand: true, dualWield: () => false,
		ranged: [R.wand], shield: false,
	},
	{
		key: 'shadow_priest', label: 'Shadow Priest', className: 'Priest', cls: Class.ClassPriest, role: 'caster',
		rotations: [{ label: 'Default', json: shadowApl }],
		spec: () => ({ oneofKind: 'shadowPriest', shadowPriest: ShadowPriest.create({ options: ShadowPriest_Options.create({}) }) }),
		armor: () => Armor.Cloth,
		weapons: [W.dagger, W.mace, W.staff, W.offhand], twoHand: true, dualWield: () => false,
		ranged: [R.wand], shield: false,
	},
	{
		key: 'balance_druid', label: 'Balance Druid', className: 'Druid', cls: Class.ClassDruid, role: 'caster',
		rotations: [{ label: 'Default', json: balanceApl }],
		spec: () => ({ oneofKind: 'balanceDruid', balanceDruid: BalanceDruid.create({ options: BalanceDruid_Options.create({}) }) }),
		armor: () => Armor.Leather,
		weapons: [W.dagger, W.fist, W.mace, W.staff, W.offhand], twoHand: true, dualWield: () => false,
		ranged: [R.idol], shield: false,
	},
	{
		key: 'feral_druid', label: 'Feral Druid', className: 'Druid', cls: Class.ClassDruid, role: 'melee',
		rotations: [{ label: 'Leveling (any level)', json: feralLevelApl }, { label: 'Level 60', json: feralApl }],
		spec: () => ({ oneofKind: 'feralDruid', feralDruid: FeralDruid.create({ options: FeralDruid_Options.create({ latencyMs: 100 }) }) }),
		armor: () => Armor.Leather,
		weapons: [W.dagger, W.fist, W.mace, W.staff, W.offhand], twoHand: true, dualWield: () => false,
		ranged: [R.idol], shield: false,
	},
	{
		key: 'retribution_paladin', label: 'Retribution Paladin', className: 'Paladin', cls: Class.ClassPaladin, role: 'melee',
		rotations: [{ label: 'Default', json: retApl }],
		spec: level => ({ oneofKind: 'retributionPaladin', retributionPaladin: RetributionPaladin.create({ options: PaladinOptions.create({
			aura: level >= 16 ? PaladinAura.RetributionAura : PaladinAura.NoPaladinAura, primarySeal: PaladinSeal.Righteousness }) }) }),
		armor: level => (level >= 40 ? Armor.Plate : Armor.Mail),
		weapons: [W.axe, W.mace, W.polearm, W.sword], twoHand: true, dualWield: () => false,
		ranged: [R.libram], shield: true,
	},
	{
		key: 'elemental_shaman', label: 'Elemental Shaman', className: 'Shaman', cls: Class.ClassShaman, role: 'caster',
		rotations: [{ label: 'Default', json: elementalApl }],
		spec: () => ({ oneofKind: 'elementalShaman', elementalShaman: ElementalShaman.create({ options: ElementalShaman_Options.create({}) }) }),
		armor: level => (level >= 40 ? Armor.Mail : Armor.Leather),
		weapons: [W.axe, W.dagger, W.fist, W.mace, W.staff, W.offhand], twoHand: true, dualWield: () => false,
		ranged: [R.totem], shield: true,
	},
	{
		key: 'enhancement_shaman', label: 'Enhancement Shaman', className: 'Shaman', cls: Class.ClassShaman, role: 'melee',
		rotations: [{ label: 'Default', json: enhancementApl }],
		spec: () => ({ oneofKind: 'enhancementShaman', enhancementShaman: EnhancementShaman.create({ options: EnhancementShaman_Options.create({}) }) }),
		armor: level => (level >= 40 ? Armor.Mail : Armor.Leather),
		weapons: [W.axe, W.dagger, W.fist, W.mace, W.staff], twoHand: true, dualWield: () => false,
		ranged: [R.totem], shield: true,
	},
];

export function rotation(def: SpecDef, index: number): APLRotation {
	return APLRotation.fromJsonString((def.rotations[index] || def.rotations[0]).json, { ignoreUnknownFields: true });
}
